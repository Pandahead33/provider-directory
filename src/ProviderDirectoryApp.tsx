import React, { ReactElement } from "react";
import AddContactForm from "./AddContactForm";
import ProviderList from "./ProviderList";
import { Contact } from "./Contact";

interface History {
  index: number;
  action: Action;
  contacts: Array<Contact>;
  deletedRecord: Contact | null | undefined;
}

enum Action {
  add = "add",
  delete = "delete",
}

interface ProviderDirectoryAppState {
  history: Array<History>;
  stepNumber: number;
}

class ProviderDirectoryApp extends React.Component<
  unknown,
  ProviderDirectoryAppState
> {
  constructor(props: unknown) {
    super(props);
    this.state = {
      history: [
        {
          index: 0,
          action: Action.add,
          deletedRecord: undefined,
          contacts: [
            {
              firstName: "Bob",
              lastName: "Smith",
              emailAddress: "bob@smith.com",
              speciality: "General Medicine",
              practiceName: "Practice Name",
            },
          ],
        },
      ],
      stepNumber: 0,
    };

    this.addRow = this.addRow.bind(this);
    this.deleteRow = this.deleteRow.bind(this);
  }

  componentDidMount(): void {
    const savedHistory = localStorage.getItem("history");
    const savedStepNumber = localStorage.getItem("stepNumber");
    if (savedHistory && savedStepNumber) {
      this.setState({
        history: JSON.parse(savedHistory),
        stepNumber: Number(savedStepNumber),
      });
    }
  }

  addRow(record: Contact): void {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const contacts = [...history[history.length - 1].contacts];
    const nextStep = this.state.stepNumber + 1;

    contacts.push({
      firstName: record.firstName,
      lastName: record.lastName,
      emailAddress: record.emailAddress,
      speciality: record.speciality || "",
      practiceName: record.practiceName || "",
    });

    history.push({
      index: nextStep,
      action: Action.add,
      contacts: contacts,
      deletedRecord: undefined,
    });

    this.setState(
      {
        history: history,
        stepNumber: nextStep,
      },
      this.updateLocalStorage
    );
  }

  deleteRow(record: Contact): void {
    const history = this.state.history.slice(0, this.state.stepNumber + 1);
    const contacts = [...history[history.length - 1].contacts];
    const nextStep = this.state.stepNumber + 1;

    if (contacts.includes(record)) {
      contacts.splice(contacts.indexOf(record), 1);

      this.setState(
        {
          history: history.concat({
            index: nextStep,
            action: Action.delete,
            contacts: contacts,
            deletedRecord: record,
          }),
          stepNumber: nextStep,
        },
        this.updateLocalStorage
      );
    }
  }

  updateLocalStorage(): void {
    localStorage.setItem("history", JSON.stringify(this.state.history));
    localStorage.setItem("stepNumber", this.state.stepNumber.toString());
  }

  jumpTo(step: number): void {
    this.setState(
      {
        stepNumber: step,
      },
      this.updateLocalStorage
    );
  }

  resetHistory(): void {
    this.setState(
      {
        history: [],
        stepNumber: 0,
      },
      this.updateLocalStorage
    );
  }

  generateActionHistoryList(
    history: Array<History>
  ): Array<JSX.Element | undefined> {
    const historyList = history.map((historyOperation) => {
      const actionMessage =
        historyOperation.action === Action.add ? "Added" : "Deleted";
      const contact =
        historyOperation.contacts[historyOperation.index] ||
        historyOperation.deletedRecord;

      if (!contact) {
        return;
      }

      const speciality = contact.speciality ? `(${contact.speciality})` : "";
      const practiceName = contact.practiceName
        ? `from ${contact.practiceName} `
        : "";
      const emailAddress = `mailto:${contact.emailAddress}`;
      const isCurrentHistory = historyOperation.index === this.state.stepNumber;
      const isPastHistory = historyOperation.index < this.state.stepNumber;
      let revertHistoryButton = (
        <span className="undo-button">
          <b>Current history</b>
        </span>
      );

      if (!isCurrentHistory) {
        revertHistoryButton = (
          <button
            className="undo-button"
            onClick={() => this.jumpTo(historyOperation.index)}
          >
            {isPastHistory
              ? "Rewind history to this point"
              : "Fast forward history to this point"}
          </button>
        );
      }

      return (
        <li key={historyOperation.index}>
          <span>
            {actionMessage}{" "}
            <a href={emailAddress}>
              {contact.firstName} {contact.lastName}
            </a>{" "}
            {practiceName} {speciality}
          </span>
          {revertHistoryButton}
        </li>
      );
    });

    return historyList.reverse();
  }

  render(): ReactElement {
    const history: Array<History> = [...this.state.history];
    const current = history[this.state.stepNumber];
    const historyList = this.generateActionHistoryList(history);

    return (
      <main>
        <ProviderList
          contacts={current.contacts}
          deleteProvider={this.deleteRow}
        />
        <div className="action-history">
          <h1>Action History</h1>
          <ol>{historyList}</ol>
        </div>
        <AddContactForm addNewProvider={this.addRow} />
      </main>
    );
  }
}

export default ProviderDirectoryApp;
