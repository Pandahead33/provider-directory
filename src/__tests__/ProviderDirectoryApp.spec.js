import React from "react"
import {render, screen, fireEvent} from '@testing-library/react'
import { create } from "react-test-renderer";

import ProviderDirectoryApp from "../ProviderDirectoryApp";

window.scrollTo = jest.fn()

test('render and matches snapshot', () => {
    expect(create(<ProviderDirectoryApp />).toJSON()).toMatchSnapshot();
});

test('add action should prepend to list on submit', () => {
    const {container} = render(<ProviderDirectoryApp />);

    fireEvent.click(screen.getByText("➕ Add New Provider"));

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');
    const emailAddressInput = screen.getByLabelText('Email Address');

    fireEvent.change(firstNameInput, { target: { value: "David"}});
    fireEvent.change(lastNameInput, { target: { value: "Hasselhoff"}});
    fireEvent.change(emailAddressInput, { target: { value: "the@dave.co"}});

    fireEvent.click(screen.getByText('➕ Add Provider'))

    expect(screen.getAllByText("Hasselhoff").length).toBeGreaterThan(0);
});

test('remove action should remove from provider list', () => {
    const {container} = render(<ProviderDirectoryApp />);

    fireEvent.click(screen.getByText("➕ Add New Provider"));

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');
    const emailAddressInput = screen.getByLabelText('Email Address');

    fireEvent.change(firstNameInput, { target: { value: "David"}});
    fireEvent.change(lastNameInput, { target: { value: "Hasselhoff"}});
    fireEvent.change(emailAddressInput, { target: { value: "the@dave.co"}});

    fireEvent.click(screen.getByText('➕ Add Provider'));

    expect(screen.getAllByText("Hasselhoff").length).toBeGreaterThan(0);

    // remove added contact
    fireEvent.click(screen.getAllByText("❌")[0]);

    expect(screen.getAllByText("Hasselhoff").length).toBe(1);
});


test('edit action should update provider in list', () => {
    const {container} = render(<ProviderDirectoryApp />);

    fireEvent.click(screen.getByText("➕ Add New Provider"));

    const firstNameInput = screen.getByLabelText('First Name');
    const lastNameInput = screen.getByLabelText('Last Name');
    const emailAddressInput = screen.getByLabelText('Email Address');

    fireEvent.change(firstNameInput, { target: { value: "David"}});
    fireEvent.change(lastNameInput, { target: { value: "Hasselhoff"}});
    fireEvent.change(emailAddressInput, { target: { value: "the@dave.co"}});

    fireEvent.click(screen.getByText('➕ Add Provider'));

    expect(screen.getAllByText("Hasselhoff").length).toBeGreaterThan(0);

    // remove added contact
    fireEvent.click(screen.getAllByText("Edit",  {selector: 'button'})[0]);
    
    const editLastNameInput = screen.getByLabelText('Last Name');
    fireEvent.change(editLastNameInput, { target: { value: "Harbringer"}});
    
    fireEvent.click(screen.getByText('Edit Provider', {selector: 'input'}));
    expect(screen.getAllByText("Harbringer").length).toBe(1);
});