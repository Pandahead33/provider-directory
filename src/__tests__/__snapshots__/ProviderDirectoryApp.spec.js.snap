// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render and matches snapshot 1`] = `
<main>
  <div
    className="provider-list"
  >
    <div
      className="search-container"
    >
      <h1>
        Provider List
      </h1>
      <input
        className="add-new-provider-button"
        onClick={[Function]}
        type="button"
        value="‚ûï Add New Provider"
      />
      <input
        className="add-new-provider-button"
        onClick={[Function]}
        type="button"
        value="‚ò†Ô∏è Delete All Providers"
      />
      <div
        className="search-bar-container"
      >
        <input
          className="search"
          onChange={[Function]}
          placeholder="Search..."
          type="search"
        />
        <label
          htmlFor="sortBy"
        >
          <div
            className="sort-by-label"
          >
            Sort By
          </div>
          <select
            id="sortBy"
            name="sortBy"
            onChange={[Function]}
            value="lastName"
          >
            <option
              value="firstName"
            >
              First Name
            </option>
            <option
              value="lastName"
            >
              Last Name
            </option>
            <option
              value="emailAddress"
            >
              Email Address
            </option>
            <option
              value="speciality"
            >
              Speciality
            </option>
            <option
              value="practiceName"
            >
              Practice Name
            </option>
          </select>
        </label>
        <div
          className="sort-direction"
        >
          <div>
            <label
              htmlFor="ascending"
            >
              <input
                checked={true}
                id="ascending"
                name="sortDirection"
                onChange={[Function]}
                type="radio"
                value="ascending"
              />
              Ascending
            </label>
          </div>
          <div>
            <label
              htmlFor="descending"
            >
              <input
                checked={false}
                id="descending"
                name="sortDirection"
                onChange={[Function]}
                type="radio"
                value="descending"
              />
              Descending
            </label>
          </div>
        </div>
      </div>
    </div>
    <span
      className="results"
    >
      1
       results
    </span>
    <div
      style={
        Object {
          "height": "0",
        }
      }
    >
      <div
        className="provider-table-heading"
      >
        <div
          className="table-header"
        >
          <p>
            First Name
          </p>
        </div>
        <div
          className="table-header"
        >
          <p>
            Last Name
          </p>
        </div>
        <div
          className="table-header"
        >
          <p>
            Email Address
          </p>
        </div>
        <div
          className="table-header"
        >
          <p>
            Speciality
          </p>
        </div>
        <div
          className="table-header"
        >
          <p>
            Practice Name
          </p>
        </div>
        <div
          className="table-header"
        >
          <p>
            Action
          </p>
        </div>
      </div>
      <div
        className="search-results-container"
      >
        <div
          className="cell"
        >
          <p>
            Bob
          </p>
        </div>
        <div
          className="cell"
        >
          <p>
            Smith
          </p>
        </div>
        <div
          className="cell"
        >
          <p>
            <a
              href="mailto:bob@smith.com"
            >
              bob@smith.com
            </a>
          </p>
        </div>
        <div
          className="cell"
        >
          <p>
            General Medicine
          </p>
        </div>
        <div
          className="cell"
        >
          <p>
            Practice Name
          </p>
        </div>
        <div
          className="cell"
        >
          <button
            className="action-button"
            onClick={[Function]}
          >
            <span
              aria-label="edit-icon"
              role="img"
            >
              üñäÔ∏è
            </span>
             
            Edit
          </button>
          <button
            className="action-button"
            onClick={[Function]}
          >
            <span
              aria-label="delete-icon"
              role="img"
            >
              ‚ùå
            </span>
             
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="action-history"
  >
    <h1>
      Action History
    </h1>
    <ol>
      <li>
        <span>
          Added
           
          <a
            href="mailto:bob@smith.com"
          >
            Bob
             
            Smith
          </a>
           
          from Practice Name 
           
          (General Medicine)
        </span>
        <span
          className="undo-button"
        >
          <b>
            Current history
          </b>
        </span>
      </li>
    </ol>
  </div>
</main>
`;
